sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,282 @@
+which requires an additional link step of the device code before being
+wrapped and integrated into the final executable.
+
+To reduce the potential size of the linked device binary, an additional host
+link step is performed to gather dependency information when static device
+libraries are being compiled.  This information is sent through the
+`clang-offload-deps` tool to generate a dependency IR file which is used
+during the device link step.
+
+There are multiple device linking steps that also occur.  The first step links
+together all of the objects and the required device libraries.  The second is
+performed including all of optional device libraries, the static device
+libraries and the dependency information that was gathered above.  This link
+step is performed with `--only-needed` to streamline the final device binary.
+
+The device libraries that are pulled in during the device link step will need
+to be controlled by the `clang-linker-wrapper`.  There are controlling options
+that are available to the user which need to be provided to the wrapper to
+understand which device libraries are not desired by the end user.
+
+### Ahead Of Time Compilation
+
+The updated offloading model will integrate the Ahead of Time (AOT) compilation
+behaviors into the clang-linker-wrapper.  The actions will typically take place
+after the device link and post link steps.
+
+Regardless of the AOT target, the flow is similar, only modifying the offline
+compiler that is used to create the target device image.  It is expected that
+the offline compiler will also use unique command lines specific to the tool
+to create the image.
+
+To support the needed option passing triggered by use of the
+`-Xsycl-target-backend` option and implied options based on the optional
+device behaviors for AOT compilations for GPU new command line interfaces
+are needed to pass along this information.
+
+| Target | Triple        | Offline Tool   | Option for Additional Args |
+|--------|---------------|----------------|----------------------------|
+| CPU    | spir64_x86_64 | opencl-aot     | `--cpu-tool-arg=<arg>`     |
+| GPU    | spir64_gen    | ocloc          | `--gen-tool-arg=<arg>`     |
+| FPGA   | spir64_fpga   | aoc/opencl-aot | `--fpga-tool-arg=<arg>`    |
+
+*Table: Ahead of Time Info*
+
+To complete the support needed for the various targets using the
+`clang-linker-wrapper` as the main interface, a few additional options will
+be needed to communicate from the driver to the tool.  Further details of usage
+are given further below.
+
+| Option Name                  | Purpose                                      |
+|------------------------------|----------------------------------------------|
+| `--fpga-tool-deps=<arg>`     | Comma separated list of dependency files used for FPGA hardware compiles using `aoc` |
+| `--parallel-link-sycl=<arg>` | Provide the number of parallel jobs that will be used when processing with `llvm-foreach` |
```

Comment:
I'm unresolving this conversation: it would actually be great to re-use existing option which specifies "amount" of parallelism which can be used by the tool. There later could be other places which may benefit from parallelization: thinLTO, for example. Also, I assume that at some point we will inline some tools into `clang-linker-wrapper` which would allow to get rid of `llvm-foreach` as well.

I'm not saying that we should use `wrapper-jobs` right away if there are semantic differences which I'm missing, but we should at least leave a note/`TODO` here that we should aim for re-using existing upstream option instead of introducing what seems to be a clone of it.